writer.go

```diff
@@ -0,0 +1,482 @@
+			if fval.IsNil() {
+				continue
+			}
+
+			skip := true
+			invalid := field.t.BaseType().Invalid()
+			for i := 0; i < fval.Len(); i++ {
+				if fval.Interface() != invalid {
+					skip = false
+					break
+				}
+			}
+			if skip {
+				continue
+			}
+		} else if fval.Interface() == allInvalid.Field(i).Interface() {
+			continue
+		}
+
+		// FIXME: No message can exceed 255 bytes
+		def.fields = append(def.fields, field)
+	}
+
+	return def
+}
+
+func (e *encoder) writeDefMesg(def *encodeMesgDef) error {
+	hdr := (1 << 6) | def.localMesgNum&0xF
+	err := binary.Write(e.w, e.arch, hdr)
+	if err != nil {
+		return err
+	}
+
+	err = binary.Write(e.w, e.arch, byte(0))
+	if err != nil {
+		return err
+	}
+
+	switch e.arch {
+	case binary.LittleEndian:
+		err = binary.Write(e.w, e.arch, byte(0))
+	case binary.BigEndian:
+		err = binary.Write(e.w, e.arch, byte(1))
+	}
+	if err != nil {
+		return err
+	}
+
+	err = binary.Write(e.w, e.arch, def.globalMesgNum)
+	if err != nil {
+		return err
+	}
+
+	err = binary.Write(e.w, e.arch, byte(len(def.fields)))
+	if err != nil {
+		return err
+	}
+
+	for _, f := range def.fields {
+		fdef := fieldDef{
+			num:   f.num,
+			size:  byte(f.t.BaseType().Size()),
+			btype: f.t.BaseType(),
+		}
+		if fdef.btype == types.BaseString {
+			fdef.size = f.length
+		} else if f.t.Array() {
+			fdef.size = fdef.size * f.length
+		}
+
+		err := binary.Write(e.w, e.arch, fdef)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (e *encoder) encodeDefAndDataMesg(mesg reflect.Value) error {
+	mesg = reflect.Indirect(mesg)
+	if !mesg.IsValid() {
+		return nil
+	}
+
+	// We'll always just use local ID 0, for simplicity
+	// We know the full file contents up-front, so no need to interleave
+	def := getEncodeMesgDef(mesg, 0)
+
+	err := e.writeDefMesg(def)
+	if err != nil {
+		return err
+	}
+
+	err = e.writeMesg(mesg, def)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+func (e *encoder) encodeFile(file reflect.Value) error {
+	for i := 0; i < file.NumField(); i++ {
+		v := file.Field(i)
+		switch v.Kind() {
+		case reflect.Struct, reflect.Ptr:
+			err := e.encodeDefAndDataMesg(reflect.Indirect(v))
+			if err != nil {
+				return err
+			}
+		case reflect.Slice:
+			var def *encodeMesgDef
+			for j := 0; j < v.Len(); j++ {
+				v2 := reflect.Indirect(v.Index(j))
+
+				if j == 0 {
+					def = getEncodeMesgDef(v2, 0)
+					err := e.writeDefMesg(def)
+					if err != nil {
+						return err
+					}
+				}
+
+				err := e.writeMesg(v2, def)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// Encode writes the given FIT file into the given Writer. file.CRC and
+// file.Header.CRC will be updated to the correct values.
+func Encode(w io.Writer, file *File, arch binary.ByteOrder) error {
+	buf := &bytes.Buffer{}
+	enc := &encoder{
+		w:    buf,
+		arch: arch,
+	}
+
+	// XXX: Is there a better way to do this with reflection?
+	var data reflect.Value
+	switch file.Type() {
```

Comment:
Would it make sense to add a default case here with an error return just in case?

Reply:
Done